yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -466,6 +470,384 @@ class FromPublisherOperator : public Flowable<T> {
+
+          auto r = elem->sync.wlock();
+          CHECK(!r->hasValue) << "failed for elem=" << elem; // sanity
+
+          elem->unlink();
+          l->pendingValue.push_back(*elem);
+        }
+        elem->request(1);
+      }
+    }
+
+    // called from MappedStreamSubscriber, recieves the R and the
+    // subscriber which generated the R
+    void drainLoop() {
+      if (drainLoopMutex_++ == 0) {
+        do {
+          drainImpl();
+        } while (drainLoopMutex_-- != 1);
+      }
+    }
+
+    void onMappedSubscriberNext(MappedStreamSubscriber* elem, R value) {
+      {
+        // `elem` may not be in a list; it may have been canceled. Push it
+        // on the withValue list and let drainLoop clear it if that's the case.
+        auto l = lists.wlock();
+        auto r = elem->sync.wlock();
+
+        CHECK(!r->hasValue) << "failed for elem=" << elem;
+        r->hasValue = true;
+        r->value = std::move(value);
+
+        elem->unlink();
+        l->withValue.push_back(*elem);
+      }
+
+      drainLoop();
+    }
+    void onMappedSubscriberTerminate(MappedStreamSubscriber* elem) {
+      liveSubscribers_--;
+
+      {
+        auto r = elem->sync.wlock();
+        if (r->onErrorEx) {
+          std::lock_guard<std::mutex> exg(onErrorExGuard_);
+          onErrorEx_ = std::move(r->onErrorEx);
+        }
+
+        if (elem->explicitlyCanceled) {
+          return;
+        }
+      }
+
+      {
+        auto l = lists.wlock();
+        CHECK(elem->is_linked());
+        elem->unlink();
+        elem->fmReference_ = nullptr;
+      }
+
+      drainLoop();
+    }
+
+    // onComplete/onError fall through to onTerminateImpl, which
+    // will call drainLoop and update the liveSubscribers_ count
+    void onCompleteImpl() final {
+    }
+    void onErrorImpl(folly::exception_wrapper ex) final {
+      std::lock_guard<std::mutex> g(onErrorExGuard_);
+      onErrorEx_ = std::move(ex);
+      clearAllSubscribers_.store(true);
+    }
+
+    void onTerminateImpl() final {
+      liveSubscribers_--;
+      drainLoop();
+    }
+
+    void request(int64_t n) override {
+      if((n + requested_) < requested_) {
+        requested_ = std::numeric_limits<int64_t>::max();
+      }
+      else {
+        requested_ += n;
+      }
+      drainLoop();
+    }
+
+    void cancel() override {
+      clearAllSubscribers_.store(true);
+      drainLoop();
+    }
+
+   private:
+    // buffers at most a single element of type R
+    struct MappedStreamSubscriber
+        : public BaseSubscriber<R>,
+          public boost::intrusive::list_base_hook<
+              boost::intrusive::link_mode<boost::intrusive::auto_unlink>> {
+      MappedStreamSubscriber(Reference<FMSubscription> subscription)
+          : flatMapSubscription_(std::move(subscription)) {}
+
+      void onSubscribeImpl() final {
+        {
+          auto l = flatMapSubscription_->lists.wlock();
+          DCHECK(is_in_list(*this, l->pendingValue, l)) << " this=" << this;
```

Comment:
Is it possible to just get rid of all these 'is_in_list' calls and the corresponding function definitions.
If we want to be sure, lets write more tests..

Reply:
These are invariant checks, I think it's helpful to leave them in. Else we'll be adding them back in when we're trying to debug issues related to these classes.